language/functional_tests/src/checker/matcher.rs

```diff
@@ -0,0 +1,182 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    checker::directives::Directive,
+    evaluator::{EvaluationLog, EvaluationOutput},
+};
+use aho_corasick::AhoCorasickBuilder;
+
+/// Splits the input directives into groups such that each except the last group ends with a directive with a
+/// positive match pattern.
+fn group_directives<I, D>(directives: I) -> Vec<Vec<(usize, D)>>
+where
+    D: AsRef<Directive>,
+    I: IntoIterator<Item = D>,
+{
+    let mut groups = vec![];
+    let mut buffer = vec![];
+
+    for (id, d) in directives.into_iter().enumerate() {
+        if d.as_ref().is_positive() {
+            buffer.push((id, d));
+            groups.push(buffer);
+            buffer = vec![];
+        } else {
+            buffer.push((id, d));
+        }
+    }
+    if !buffer.is_empty() {
+        groups.push(buffer);
+    }
+
+    groups
+}
+
+/// A single match consisting of the index of the log entry, the start location and the end location (in bytes).
+#[derive(Debug)]
+pub struct Match {
+    pub pat_id: usize,
+    pub entry_id: usize,
+    pub start: usize,
+    pub end: usize,
+}
+
+#[derive(Debug)]
+pub enum MatchError {
+    NegativeMatch(Match),
+    UnmatchedDirectives(Vec<usize>),
+    UnmatchedErrors(Vec<usize>),
+}
+
+#[derive(Debug)]
+pub enum MatchStatus {
+    Success,
+    Failure(Vec<MatchError>),
+}
+
+/// The result of matching the directives against the evaluation log.
+#[derive(Debug)]
+pub struct MatchResult {
+    pub status: MatchStatus,
+    pub text: Vec<String>,
+    pub matches: Vec<Match>,
+}
+
+/// Matches the directives against the evaluation log.
+pub fn match_output<I, D>(log: &EvaluationLog, directives: I) -> MatchResult
+where
+    D: AsRef<Directive>,
+    I: IntoIterator<Item = D>,
+{
+    // Convert each entry of the evaluation log into a string, which will be later matched against.
+    let text: Vec<_> = log
+        .outputs
+        .iter()
+        .map(|output| format!("{:?}", output))
+        .collect();
+
+    // Split directives into groups and build an Aho-Corasick automaton for each group.
+    let groups = group_directives(directives);
+    let automata: Vec<_> = groups
+        .iter()
+        .map(|directives| {
+            AhoCorasickBuilder::new()
```

Comment:
You are going to need a lot of comments here... 
Maybe split this out into a helper so you can explain what's going on. Just citing Aho-Corasick is a bit opaque

Reply:
In other words, this is an algorithm for matching, so you should having something called 'matcher' or 'dictionary' and the algorithm can be hidden inside of that

Reply:
Yeah I'll refactor the code here & add a few paragraphs explaining how the algorithm works.